### DISCUSSION TITLE
Trouble restoring settings for streamed installed apps

### DISCUSSION THREAD

**Discussion by webenhanced** [mrrfv/open-android-backup#51] [Q&A]
I have been testing this script on a Pixel 5 phone with CalyxOS 4.6.0 (android 13) and cannot get the data or settings connected to the two apps I am testing. I factory reset the phone then install the two apps and configure them.

Both apps are from fdroid. They are:
- org.secuso.privacyfriendlysudoku
- com.fsck.k9

I commented out the companion app parts of backup.sh, backup_func.sh, and restore_func.sh because it fails on the contacts and sms. This may be due to contacts being synced to my own nextcloud instance, or maybe because there is 400 odd contacts there, not sure. In any case, I am not concerned about contacts because I just use davx and sync to the nextcloud server. As for sms and call logs, I found a good little app to export and import them to/from Documents and that export is backed up by this script. The app I mention is https://play.google.com/store/apps/details?id=com.riteshsahu.SMSBackupRestore.

Once the restore is completed with no errors, I open the two apps (above) and they behave like a brand new install, asking for permissions and other initialisations as if there was no data.

I noticed that there was no "Storage" location for the sudoku app so not sure where it keeps its settings and highscore data.

As for the K9 email client, there is a Storage location under Android/data/com.fsck.k9 but no files in there, nor are there any there when the app is installed and configured.

Any help would be appreciated.

I am using a PC with Fedora 36 Linux with latest updates etc. I am also using USB cable for the testing. It is the same cable that I used to flash the CalyxOS rom onto the device.

Thanks, Graham

**Comment by webenhanced** [comment]
I notice that the data for K9 email client seems to live in a directory outside of /storage/. On my daily driver phone (Pixel2XL with LineageOS), with "adb root" then "adb shell", I see the data in the filesystem here:-
/data/user/0/com.fsck.k9/
drwx------ 10 u0_a184        u0_a184        4096 2022-08-02 10:24 com.fsck.k9

All directories in /data/user/0/ have unique UIDs and GIDs, so I am guessing that user permissions would have to be set on that folder.

On the CalyxOS Pixel 5, adb root gives permission denied. Bugger.

**Comment by mrrfv** [comment]
Unfortunately, Linux Android Backup is unable to back up app data because Android tightly controls which apps can access the data directory - you need either root access or a privileged system app to do that, and this project is advertised as working without root. I think it is possible to add this functionality using [hooks](https://github.com/mrrfv/linux-android-backup#hooks) ([example](https://github.com/mrrfv/linux-android-backup/blob/master/hooks.example.sh)) though.

Most apps nowadays sync to the cloud or give you an option to export their data to the internal storage (like Signal), so you can do that as a workaround.

> it fails on the contacts and sms. This may be due to contacts being synced to my own nextcloud instance, or maybe because there is 400 odd contacts there, not sure.

Could you explain where does it fail (i.e. does the companion app lag/crash, or is it the script that's showing error messages)? I'd like to fix this.

**Reply by webenhanced** [reply]
I'll uncomment the companion app and test it again. Will report back here with as much info as I can gather.

**Reply by mrrfv** [reply]
Thanks. If the problem lies within the companion app, you can open the app's source code in Android Studio (`companion_app` directory, not included in the release bundles; a working Flutter install on your computer is required) and start it in debug mode to see detailed info btw.

**Reply by webenhanced** [reply]
The process looked like it ran OK on my test rig. So no SMS problem this time. But only 284 contacts were exported out of 414.

Here's a pastbin of the logcat around that time. It probably has plenty of extra stuff in there not applicable to your script and app, but I included it all in case you need it.
https://pastebin.com/UgWZWsVG

Here is the on screen response of the session:
Test using P5

Load davx and sync contacts from my nextcloud instance.

Receive a couple of SMS messages.

Receive calls for call log.

Use latest LAB from github and copy to:
	- /home/graham/DATA/AOSP/LAB-latest
Run backup.sh and report errors and results.
/home/graham/DATA/AOSP/LAB-latest/backup.sh
	(wired / tar / nohook / backup / .)
	- loads companion app - file permissions granted on phone
	- Phone says data exported - please continue on computer

After q&a start:
Starting: Intent { cmp=mrrfv.backup.companion/.MainActivity }
The companion app has been opened on your device. Please press the 'Export Data' button - this will export contacts to the internal storage, allowing this script to backup them. Press Enter to continue.

Attempting to uninstall companion app.
Exporting apps.
Backing up app: com.fsck.k9
7.60MiB 0:00:00 [19.7MiB/s] [  <=>                                                                                                                             ]
Backing up app: org.secuso.privacyfriendlysudoku
3.18MiB 0:00:00 [13.2MiB/s] [  <=>                                                                                                                             ]
Backing up app: at.bitfire.davdroid
10.8MiB 0:00:00 [14.8MiB/s] [  <=>                                                                                                                             ]
Exporting contacts (as vCard) and SMS messages (as CSV).
 332KiB 0:00:00 [1.16MiB/s] [  <=>                                                                                                                             ]
Removing temporary files created by the companion app.
Exporting internal storage - this will take a while.
42.0KiB 0:00:00 [ 764KiB/s] [  <=>                                                                                                                             ]
Exporting call logs.

All required data has been copied from your device and it can now be unplugged.

Compressing & encrypting data - this will take a while.

7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_AU.UTF-8,Utf16=on,HugeFiles=on,64 bits,12 CPUs Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (906EA),ASM,AES-NI)

Scanning the drive:
28 folders, 302 files, 22779286 bytes (22 MiB)

Creating archive: ./linux-android-backup-02-28-2023-09-51-57.7z

(bunch of backup-tmp/.....)

Files read from disk: 297
Archive size: 16058874 bytes (16 MiB)
Everything is Ok
Cleaning up after backup/restore...
Using the 'Fast' data erase mode.

**Reply by webenhanced** [reply]
I will run the backup on the daily driver again and catch the logs for that when it fails. There must be something different over there.
The P5 test rig was factory reset then a couple of apps installed, then synced the contacts with cloud then backup.sh.
